From: Baolin Wang <baolin.wang@linux.alibaba.com>
Subject: mm: rmap: skip batched unmapping for UFFD vmas
Date: Sat, 17 Jan 2026 00:26:52 +0800

As Dev reported[1], it's not ready to support batched unmapping for uffd
case.  Let's still fallback to per-page unmapping for the uffd case.

Link: https://lkml.kernel.org/r/20260116162652.176054-1-baolin.wang@linux.alibaba.com
Signed-off-by: Baolin Wang <baolin.wang@linux.alibaba.com>
Reported-by: Dev Jain <dev.jain@arm.com>
Closes: https://lore.kernel.org/linux-mm/20260116082721.275178-1-dev.jain@arm.com/ [1]
Suggested-by: Barry Song <baohua@kernel.org>
Cc: Catalin Marinas <catalin.marinas@arm.com>
Cc: David Hildenbrand <david@kernel.org>
Cc: Jann Horn <jannh@google.com>
Cc: Liam Howlett <liam.howlett@oracle.com>
Cc: Lorenzo Stoakes <lorenzo.stoakes@oracle.com>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: Michal Hocko <mhocko@suse.com>
Cc: Mike Rapoport <rppt@kernel.org>
Cc: Rik van Riel <riel@surriel.com>
Cc: Ryan Roberts <ryan.roberts@arm.com>
Cc: Suren Baghdasaryan <surenb@google.com>
Cc: Vlastimil Babka <vbabka@suse.cz>
Cc: Will Deacon <will@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/rmap.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/mm/rmap.c~mm-rmap-support-batched-unmapping-for-file-large-folios-fix
+++ a/mm/rmap.c
@@ -1953,6 +1953,9 @@ static inline unsigned int folio_unmap_p
 	if (pte_unused(pte))
 		return 1;
 
+	if (userfaultfd_wp(vma))
+		return 1;
+
 	return folio_pte_batch(folio, pvmw->pte, pte, max_nr);
 }
 
_
